### PR TITLE
[flutter_tools] refactor the IOSDevicePortForwarder and move tests out of devices_test.dart

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

(Note: I tried the lib only changes, but those need test changes too .. so it got pretty big. Trying vertical cuts instead of horizontal).

Updates the IOSDevicePortForwarder to no longer depend on context, or on an IOSDevice instance. Instead, it receives all necessary configuration through the constructor.

Moves the IOSDevicePortForwarder to a separate file.

### CODE REVIEW

File: packages/flutter_tools/lib/src/ios/devices.dart [hunk 1]
```diff
@@ -364,7 +365,13 @@ class IOSDevice extends Device {
   }
 
   @override
-  DevicePortForwarder get portForwarder => _portForwarder ??= IOSDevicePortForwarder(this);
+ DevicePortForwarder get portForwarder => _portForwarder ??= IOSDevicePortForwarder(
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/ios/devices.dart]: This is not referenced in google3

File: packages/flutter_tools/lib/src/ios/devices.dart [hunk 2]
```diff
@@ -582,20 +589,56 @@ class IOSDeviceLogReader extends DeviceLogReader {
...
-  final List<ForwardedPort> _forwardedPorts;
+  final ProcessUtils _processUtils;
+  final Logger _logger;
+  final MapEntry<String, String> _dyLdLibEntry;
+  final String _id;
+  final String _iproxyPath;
 
   @override
-  List<ForwardedPort> get forwardedPorts => _forwardedPorts;
+  List<ForwardedPort> forwardedPorts = <ForwardedPort>[];
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/ios/devices.dart]: This is semantically equivalent, just shorter

File: packages/flutter_tools/lib/src/ios/devices.dart [hunk 3]
```diff
@@ -582,20 +589,56 @@ class IOSDeviceLogReader extends DeviceLogReader {
...
+      dyLdLibEntry: const MapEntry<String, String>(
+        'DYLD_LIBRARY_PATH', '/path/to/libs',
+      ),
+    );
+  }
 
-  final List<ForwardedPort> _forwardedPorts;
+  final ProcessUtils _processUtils;
+  final Logger _logger;
+  final MapEntry<String, String> _dyLdLibEntry;
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/ios/devices.dart]: non-trivial change: I inject the dyLdLibEntry and iproxy paths.

File: packages/flutter_tools/test/src/fake_process_manager.dart [hunk 4]
```diff
@@ -139,8 +137,12 @@ class _FakeProcess implements Process {
         }
         return _exitCode;
       }),
-      stderr = Stream<List<int>>.value(utf8.encode(_stderr)),
-      stdout = Stream<List<int>>.value(utf8.encode(_stdout));
+      stderr = _stderr == null
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/src/fake_process_manager.dart]: We needed a way to differentiate the stream being empty versus containing an empty string.

File: packages/flutter_tools/test/general.shard/ios/ios_device_port_forwarder_test.dart [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+const Map<String, String> kDyLdLibEntry = <String, String>{
+  'DYLD_LIBRARY_PATH': '/path/to/libs',
+};
+
+void main() {
+  // By default, the .forward() method will try every port between 1024
+  // and 65535; this test verifies we are killing iproxy processes when
+  // we timeout on a port
+  testWithoutContext('IOSDevicePortForwarder.forward will kill iproxy processes before invoking a second', () async {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/ios/ios_device_port_forwarder_test.dart]: We only have a single IOSDevicePortForwarder test :(

File: packages/flutter_tools/lib/src/ios/devices.dart (hunk 2 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/devices.dart]: This is a valid override for `List<ForwardedPort> get forwardedPorts`? TIL.

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/ios/devices.dart]: Dart doesn't actually have fields, so:

```
int get a => _a;
set a(int value) => _a = a;
```
is equivalent to

```
int a;
```

And
```
int get a => _a;
```

Is equivalent to a final field like above.

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/ios/devices.dart]: (edited, code typo)

**Review by zanderso — APPROVED** [review]

File: packages/flutter_tools/lib/src/ios/devices.dart (hunk 2 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/devices.dart]: TIL

File: packages/flutter_tools/test/src/fake_process_manager.dart (hunk 4 above)
Comment (christopherfujino) [reply packages/flutter_tools/test/src/fake_process_manager.dart]: what does `utf8.encode('')` return?

**Review by christopherfujino — APPROVED** [review]
LGTM

File: packages/flutter_tools/test/src/fake_process_manager.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/src/fake_process_manager.dart]: EmptyList, but the Stream constructor leads to a non-empty stream with a single empty list emitted on it.

**Pull request merged**